### ISSUE TITLE
[Question] Briefly describe your problems here

### ISSUE THREAD

**Issue by KirubakaranSP** [hugegraph/hugegraph#1502]
Read time out : Socket time out exception when executing gremlin query via java

## Environment ( 环境信息 - 必填 )
- **Server Version**: v0.11.x (refer [here](https://hugegraph.github.io/hugegraph-doc/clients/restful-api/other.html))
- **Backend**: Cassandra 3.x, x nodes, HDD or SSD
- **OS**: Ubuntu,128 GB RAM
- **Data Size**: 10 million vertex 30 million edges

I am executing gremlin query  which includes multiple traversals .It results in Socket time out exception .I am executing the query through JAVA API.
**Configurations** : 
Macine RAM : 128 GB
 RAM configured for this process : 48GB

**Sample code** : 
GremlinRequest gr =new  GremlinRequest(queryString);
hugeClient().gremlin().execute(gr);

**Query**  :  
g.V().or((has('METHOD_CALL', 'CLASS_NAME', 'javax/servlet/http/HttpServletRequest')).or(has('METHOD_CALL', 'METHOD_NAME', 'getAttribute'),has('METHOD_CALL', 'METHOD_NAME', 'getParameter')),(has('METHOD_CALL', 'CLASS_NAME', 'javax/servlet/ServletRequest')).or(has('METHOD_CALL', 'METHOD_NAME', 'getAttribute'),has('METHOD_CALL', 'METHOD_NAME', 'getParameter')))..repeat(outE().otherV(). simplePath()).until(and(has('METHOD_CALL', 'CLASS_NAME', 'java/io/PrintWriter').has('METHOD_CALL', 'METHOD_NAME', 'println').repeat(inE().otherV().simplePath()) .until( and(has('METHOD_CALL', 'CLASS_NAME', 'javax/servlet/http/HttpServletResponse').has('METHOD_CALL', 'METHOD_NAME', 'getWriter'))))).path()

Am i missing any configurations here or is there anything problem with the query?
Kindly check and provide a fix for this.

**Comment by imbajin** [comment]
Some advices:
1. figure out how many records your query got for once (like use `g.V().xxx().count()` to estimate)
2. if `count()` also return `timeout exception`, u can change the `socket timeout` configs in 2 modules:
   - first in `rest-server.properties` and `gremlin-server.yaml` restart **graph server** (refer the [doc](https://hugegraph.github.io/hugegraph-doc/config/config-option.html), default is 30s)
   - then set timeout params for **HugeClient** (refer [here](https://hugegraph.github.io/hugegraph-doc/clients/hugegraph-client.html))
3. use direct REST-API sender (like postman or [talend](https://www.talend.com/)) to send request (if HugeClient still get timeout)

**Comment by KirubakaranSP** [comment]
Records are more and socket timeout even occurs for the count query
sometimes.I forgot to mention the configurations,the below  configurations
are already configured
task.wait_timeout=600000
batch.max_write_ratio=90
batch.max_write_threads=100
batch.max_vertices_per_batch=20000
batch.max_edges_per_batch=20000
restserver.request_timeout=-1
restserver.connection_idle_timeout=-1
gremlinserver.timeout=1000000
oltp.concurrent_threads=20
oltp.concurrent_depth=20
scriptEvaluationTimeout: 5000000
configTImeout in Hugeclient is -1
Is there anything need to be configured for Gremlin traversal query.When
query involves more traversal,then the exception occurs.Kindly update me

**Comment by imbajin** [comment]
seems your graph sever's config are OK.

do you try direct REST-api sender to send gremlin? [refer [here](https://hugegraph.github.io/hugegraph-doc/clients/restful-api/gremlin.html)] (if not, try first, and u can verify the configs has **taken effect** by using the medium-time cost query. (e.g `g.V().count(10000000)`)

then found a wrong config:
> configTImeout in Hugeclient is -1

refer to jeresy's official [doc](https://eclipse-ee4j.github.io/jersey.github.io/apidocs/latest/jersey/org/glassfish/jersey/client/ClientProperties.html#CONNECT_TIMEOUT), seems u should set `0` for `configTImeout()` rather than `-1` (we will add desc for it later)

**Comment by KirubakaranSP** [comment]
If config time is 0 , then default time out 20 will be set in hugegraph.Can
you please check the code

 *public* HugeClientBuilder configTimeout(*int* timeout) {

        *if* (timeout == 0) {

            timeout = *DEFAULT_TIMEOUT*;

        }

        *this*.timeout = timeout;

        *return* *this*;

    }

**Comment by imbajin** [comment]
Thanks for reminding and correcting the description 😢 

and any update for the direct `rest-api` request?

BTW, there was another way to get long-time tasks without changing the configs, use `async-gremlin` way ( refer [here](https://hugegraph.github.io/hugegraph-doc/clients/restful-api/gremlin.html#713-%E5%90%91hugegraphserver%E5%8F%91%E9%80%81gremlin%E8%AF%AD%E5%8F%A5%EF%BC%88post%EF%BC%89%EF%BC%8C%E5%BC%82%E6%AD%A5%E6%89%A7%E8%A1%8C) )

And u can also use it in HugeClient (like use `GremlinManager.executeAsTask()`)

**Comment by KirubakaranSP** [comment]
I've tried using postman but it takes too much time to process a query.Here
is my use case
Am iterating over millions of vertices and edges by using gremlin's repeat
statement.Can you please verify the below query?

g.V().has('Vertex1', 'PROP_NAME1', 'PROP_VALUE1').has('Vertex1',
'PROP_NAME2', 'PROP_VALUE2').repeat(outE().otherV().
simplePath()).until(and(has('Vertex1', 'PROP_NAME1',
'PROP_VALUE3').has('METHOD_CALL', 'PROP_NAME2',
'PROP_VALUE4))).repeat(inE().otherV().simplePath())
.until(and(has('Vertex1', 'PROP_NAME1', 'PROP_VALUE5').has('Vertex1',
'PROP_NAME2', 'PROP_VALUE6'))).path()

Part of the query is providing results(say for first repeat).KIndly correct
me if I did anything wrong with the query?

And regd async gremlin how can i get the result  of the task and how did i
know that task is completed?Can you provide me with any or example for this?

P.S : executing g.V().count() from postman executed successfully.

**Comment by imbajin** [comment]
for the query, u should describe the needs first (like "I want to query the neighbor start from `tom` and depth = 3"), then we could ensure if the query has better solution

then for async gremlin, after u get a response with the taskID, u could reqeust [task-api](https://hugegraph.github.io/hugegraph-doc/clients/restful-api/task.html) to get the query's progress & result, also refer the [doc](https://hugegraph.github.io/hugegraph-doc/clients/restful-api/task.html)

![image](https://user-images.githubusercontent.com/17706099/122401770-7a1ed700-cfaf-11eb-8aa4-c8e26f906e8d.png)


last for the huge-client, u can refer the test code [here](https://github.com/hugegraph/hugegraph-client/blob/bb6549230eb883f95be39155a2f23f15a912503c/src/test/java/com/baidu/hugegraph/api/GremlinApiTest.java#L88)

**Comment by KirubakaranSP** [comment]
I apologize for not explaining the query.Here is the use case,
I want to query all outgoing paths from A to B then the incoming paths from
B to C (there is no depth specified ,need to check all the possible
traversals).Kindly check the attached image.
And regd REST TASK API, I can't see results in response.Can you please
clarify?

**Comment by imbajin** [comment]
if u wanna get `all paths` between 2 or more vertices, u should limit depth (or elements amount) first, like limit the `repeat()` or `time()`. 

Based on your needs, u should try to use the more efficient and simple way:
 - use [Paths GET API](3213-paths-（get，基础版）) for simple query
 - use [Paths POST API](https://hugegraph.github.io/hugegraph-doc/clients/restful-api/traverser.html#3214-paths-%EF%BC%88post%EF%BC%8C%E9%AB%98%E7%BA%A7%E7%89%88%EF%BC%89) (or [Customized Paths](url) API) for complex query

And the default `max_depath` is 10 (u can easily modify it)

---
if u cannot get results for async task, u need paste your screenshot first.. (at least with the response `json` result)

![image](https://user-images.githubusercontent.com/17706099/122727966-33c8c100-d2aa-11eb-90cf-20d1b83a9726.png)

**Comment by KirubakaranSP** [comment]
Hi,
I've tried using PATHS POST API,but it results in some errors.I have
attached the screenshot for your reference.
Use Case :
I am getting all the possible routes from airports CAA to MAA.
Sample JSON :
{"sources": {
"label": "ROUTE",
"properties": {
"AIRPORT_NAME": "CAA",
"FLIGHT_TYPE": "Cargo"
}
},"targets": {
"label": "ROUTE",
"properties": {
"AIRPORT_NAME": "MAA",
"FLIGHT_TYPE": "Cargo"
}
},"step": {
"direction": "OUT"
},
"max_depth": 100000000,
"capacity": 100000000,
"limit": 10000000,
"with_vertex": true
}

**Comment by imbajin** [comment]
![image](https://user-images.githubusercontent.com/17706099/122903591-21ba5180-d382-11eb-8ea7-005052b90d58.png)

A little confused about your "screenshot", could u please check where it is...

And for Path API, do u try Path-Get first? Is it OK? 

> but it(path post api) results in some errors

and add more info about the **error** (maybe u should use github-app or website way to paste img?)

**Comment by KirubakaranSP** [comment]
Sorry attachement was not added from Gmail.
Yeah ,GET API works fine for traversal, but POST results in error.In my use-case only label name properties will be known ,id will not be known.
![POSTAPI](https://user-images.githubusercontent.com/40020588/122908024-507bfc00-d371-11eb-8369-b73f679d5958.png)

**Comment by imbajin** [comment]
seems the error msg tell u the reason:
![image](https://user-images.githubusercontent.com/17706099/122953233-234f3e00-d3b1-11eb-9fc5-cd74de9e12c7.png)

please reduce your source vertices amount <= 10.

if u want to get all paths from multiple vertices at once, u could send them separately.

PS: The Paths API & other api can also be used in HugeClient, after test it by postman, u could wrapper it with client

**Comment by KirubakaranSP** [comment]
In my case ,source vertices will be huge count say count will be around
million .I want to match these conditions with million of vertices

**Comment by imbajin** [comment]
Afraid not.

the performance of rest-api (likes paths-post) usually much more better than the gremlin query (for same needs) and u can adjust the path-api code by yourself (like expand the limit of source vertices, or wrapper it)

and the gremlin query also limit the vertices params (<=250), if the params of vertices around million... seems I don't know if there is a better way. (maybe u could ask the author of tinkerpop)

**Comment by KirubakaranSP** [comment]
Thanks for you response .As you methioed, rest-api is better than executinng gremlin query in java .Regd gremlin query you mentionned "the gremlin query also limit the vertices params (<=250)" , is that meant that number of vertices in the traversal must be less than 250?

**Comment by imbajin** [comment]
> is that meant that number of vertices in the traversal must be less than 250?

To be precise, not in the `traversal` process, but in each gremlin step(like `E()`, `V()`, `has()`), and it means that `gremlin` itself is not suitable for `OLAP` query. (no limit for depth or paths amount)

**Comment by KirubakaranSP** [comment]
I apologize for the delayed response.I have some doubts regd data insertion  and query execution in hugegraph .Can I raise new issue or continue with this ?

**Comment by imbajin** [comment]
1. maybe lack some context (like what kind of query?) and the same query may hit the cache (the key of cache is queryId), u can close the cache or limit the capacity to 1 to ensure
2. the simplest & official way is to use `hugegraph-loader`, like with backend rocksdb (ssd disk), usually could write vertex/edge with **300k/s**, u could try it to test
3. if u cannot use loader because u have to custom data in huge-client, u can use spark to load elements in parallel to speed up (and at last, if your data amount > 50 billion, u could use bulk-load mode)

**Comment by KirubakaranSP** [comment]
the query is repeat traversal query which will traverse over millions of edges and vertices.I've mentioned the sample query above.
g.V().has('ROUTE', 'AIRPORT_NAME', 'CAA').has('ROUTE', 'FLIGHT_TYPE', 'Cargo').repeat(outE().otherV(). simplePath()).until(and(has('ROUTE', 'AIRPORT_NAME', 'MAA').has('ROUTE', 'FLIGHT_TYPE', 'Cargo'))).repeat(inE().otherV().simplePath()) .until(and(has('ROUTE', 'AIRPORT_NAME', 'XIA').has('ROUTE', 'FLIGHT_TYPE', 'Cargo'))).path()

**Comment by imbajin** [comment]
u can close default elements cache first, it will use a lot of memory

then could u print a fire flame graph by using [profile tool](https://github.com/jvm-profiling-tools/async-profiler), it will tell us what method occupy cpu resource

**Comment by KirubakaranSP** [comment]
Can u guide me how to close default cache?
And regarding memory ,data insertion (edge data insertion)  and some repeat traversal queries occupied lot of memory

**Comment by imbajin** [comment]
> Can u guide me how to close default cache?
> And regarding memory ,data insertion (edge data insertion) and some repeat traversal queries occupied lot of memory

refer [doc](https://hugegraph.github.io/hugegraph-doc/config/config-option.html), search `cache` for keyword, and set vertex/edge cache to 1 (not include schema cache)

**Comment by KirubakaranSP** [comment]
In this [link](https://hugegraph.github.io/hugegraph-doc/quickstart/hugegraph-spark.html) , it explains how to analyse using spark.Can u tell how to use spark to load bulk data ? Can we use JSON  format like we use to insert in hugegraph?
